src/game_api/prng.cpp

```diff
@@ -0,0 +1,33 @@
+#include "prng.hpp"
+
+#include "state.hpp"
+
+PRNG& PRNG::get()
+{
+	const auto& state = State::get();
+	static PRNG* prng = (PRNG*)((size_t)state.ptr() - 0xb0);
+	return *prng;
+}
+PRNG& PRNG::get_local()
+{
+	const auto& state = State::get();
+	static PRNG* prng = (PRNG*)((size_t)state.ptr_local() - 0xb0);
+	return *prng;
+}
+
+std::size_t PRNG::random_index(std::size_t size, PRNG_CLASS type)
+{
+	prng_pair& pair = pairs[type];
+
+	const std::uint64_t lower = pair.first;
```

Comment:
@zappatic This code seems to generate almost identical assembly as the game contains, only with different instruction ordering. But I thought that it shouldn't actually matter, right? As long as we do something deterministic here and touch the same pairs as the corresponding code in the game would it should all be fine.

Reply:
Yes, the order shouldn't really matter, because the two things happening (advancing the pair and calculating the random number) are separate from each other, as the random number generation happens on the already existing pair value, not the advance.
I assume the Ghidra C++ equivalent of the advance is correct. Unless there is some very specific math behind them, like using specific primes or something, I don't believe it really matters which 64 bit value you end up after the advance anyway, as long as it is big enough. Looks good and deterministic indeed.